pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -1196,35 +1196,48 @@ private CompletableFuture<Void> delete(boolean failIfHasSubscriptions,
                                     unfenceTopicToResume();
                                     deleteFuture.completeExceptionally(e);
                                 } else {
-                                    ledger.asyncDelete(new AsyncCallbacks.DeleteLedgerCallback() {
-                                        @Override
-                                        public void deleteLedgerComplete(Object ctx) {
-                                            brokerService.removeTopicFromCache(topic);
-
-                                            dispatchRateLimiter.ifPresent(DispatchRateLimiter::close);
-
-                                            subscribeRateLimiter.ifPresent(SubscribeRateLimiter::close);
-
-                                            unregisterTopicPolicyListener();
-
-                                            log.info("[{}] Topic deleted", topic);
-                                            deleteFuture.complete(null);
-                                        }
+                                    // Truncate to ensure the offloaded data is not orphaned.
+                                    // Also ensures the BK ledgers are deleted and not just scheduled for deletion
+                                    CompletableFuture<Void> truncateFuture = ledger.asyncTruncate();
+                                    truncateFuture.whenComplete((ignore, exc) -> {
```

Comment:
How about move this logic to ledger.asyncDelete(). It can cover more situation.

Reply:
@horizonzy moved it into ManagedLedgerImpl